File: src/main/java/chapter5/FlipExample.java
package chapter5;

import common.CommonUtils;
import common.Log;
import common.Shape;
import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;


public class FlipExample {

    public void marbleDiagram() {
        String[] objs = {"1-S", "2-T", "3-P"};
        Observable<String> source = Observable.fromArray(objs)
                .doOnNext(data -> Log.it("Original data = " + data))
                .subscribeOn(Schedulers.newThread())
                .observeOn(Schedulers.newThread())
                .map(Shape::flip);
        source.subscribe(Log::it);
        CommonUtils.sleep(500);
    }

    public void observeOnRemoved() {
        String[] objs = {"1-S", "2-T", "3-P"};
        Observable<String> source = Observable.fromArray(objs)
                .doOnNext(data -> Log.it("Original data = " + data))
                .subscribeOn(Schedulers.newThread())
                // .observeOn(Schedulers.newThread())
                .map(Shape::flip);
        source.subscribe(Log::it);
        CommonUtils.sleep(500);
    }

    public static void main(String[] args) {
        FlipExample demo = new FlipExample();
        demo.marbleDiagram();
        demo.observeOnRemoved();
    }

}

File: src/main/kotlin/chapter5/FlipExample_.kt
package chapter5

import common.CommonUtils_
import common.Log_
import common.Shape_
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers


class FlipExample_ {

    fun marbleDiagram() {
        val objs = arrayOf("1-S", "2-T", "3-P")
        val source = Observable.fromArray(*objs)
                .doOnNext { Log_.it("Original data = $it") }
                .subscribeOn(Schedulers.newThread())
                .observeOn(Schedulers.newThread())
                .map { Shape_.flip(it) }
        source.subscribe { Log_.it(it) }
        CommonUtils_.sleep(500)
    }

    fun observeOnRemoved() {
        val objs = arrayOf("1-S", "2-T", "3-P")
        val source = Observable.fromArray(*objs)
                .doOnNext { Log_.it("Original data = $it") }
                .subscribeOn(Schedulers.newThread())
                // .observeOn(Schedulers.newThread())
                .map { Shape_.flip(it) }
        source.subscribe { Log_.it(it) }
        CommonUtils_.sleep(500)
    }

}


fun main(args: Array<String>) {
    val demo = FlipExample_()
    demo.marbleDiagram()
    demo.observeOnRemoved()
}
